code/datums/components/bump_attack.dm

```diff
@@ -0,0 +1,57 @@
+/datum/component/bump_attack
+    var/active = FALSE
+    var/datum/action/bump_attack_toggle/toggle_action
+
+/datum/component/bump_attack/Initialize()
+	. = ..()
+	if(!ismovableatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	if(isliving(parent))
+		toggle_action = new()
+		RegisterSignal(toggle_action, COMSIG_ACTION_TRIGGER, .proc/bump_attack_toggle)
+		toggle_action.give_action(parent)
+		toggle_action.update_button_icon(active)
+	else
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/bump_attack/proc/bump_attack_toggle(datum/source, atom/target)
+	var/mob/living/bumper = parent
+
+	active = !active
+	to_chat(bumper, "<span class='notice'>You will now [active ? "attack" : "push"] enemies who are in your way.</span>")
+
+	if(active)
+		RegisterSignal(bumper, COMSIG_MOVABLE_BUMP, .proc/bump_action)
+	else
+		UnregisterSignal(bumper, COMSIG_MOVABLE_BUMP)
+	toggle_action.update_button_icon(active)
+
+/datum/component/bump_attack/proc/bump_action(datum/source, atom/target)
+    if(isliving(parent))
```

Comment:
Sanity check not needed. Parent is always living.

Reply:
Same here, it's expected this component to be easily adaptable to non-living.
I'd move the `to_chat(bumper, "<span class='notice'>You will now [active ? "attack" : "push"] enemies who are in your way.</span>")` to this proc instead, and remove the need to typecast parent in `/bump_attack_toggle(`

Reply:
nemvar is right and you can register with a different proc earlier based on the signal. If you need to lock all the code in a receiving proc behind an istype(parent) you probably shouldn't have registered the proc.